db-service/lib/deep-genres.sql

```diff
@@ -0,0 +1,126 @@
+-- DEBUG => this.dbc._native.prepare(cds.utils.fs.readFileSync(__dirname + '/deep-genres.sql','utf-8')).exec([JSON.stringify(query.UPDATE.data)])
+DO (IN JSON NCLOB => ?) BEGIN
+
+  -- Extract genres with a depth of 3 (like: '$.children[*].children[*]')
+  Genres = SELECT
+    NEW.name,
+    NEW."$.NAME",
+    NEW.descr,
+    NEW."$.DESCR",
+    NEW.ID,
+    NEW."$.ID",
+    NEW.parent_ID,
+    NEW."$.PARENT_ID",
+    NEW."$.CHILDREN"
+  FROM 
+    JSON_TABLE(
+      :JSON,
+      '$' COLUMNS(
+        name NVARCHAR(1020) PATH '$.name',
+        "$.NAME" NVARCHAR(2147483647) FORMAT JSON PATH '$.name',
+        descr NVARCHAR(4000) PATH '$.descr',
+        "$.DESCR" NVARCHAR(2147483647) FORMAT JSON PATH '$.descr',
+        ID INT PATH '$.ID',
+        "$.ID" NVARCHAR(2147483647) FORMAT JSON PATH '$.ID',
+        parent_ID INT PATH '$.parent_ID',
+        "$.PARENT_ID" NVARCHAR(2147483647) FORMAT JSON PATH '$.parent_ID',
+        "$.CHILDREN" NVARCHAR(2147483647) FORMAT JSON PATH '$.children'
+      )
+      ERROR ON ERROR
+    ) AS NEW
+      UNION ALL
+  SELECT
+    NEW.name,
+    NEW."$.NAME",
+    NEW.descr,
+    NEW."$.DESCR",
+    NEW.ID,
+    NEW."$.ID",
+    NEW.parent_ID,
+    NEW."$.PARENT_ID",
+    NEW."$.CHILDREN"
+  FROM 
+    JSON_TABLE(
+      :JSON,
+      '$.children[*]' COLUMNS(
+        name NVARCHAR(1020) PATH '$.name',
+        "$.NAME" NVARCHAR(2147483647) FORMAT JSON PATH '$.name',
+        descr NVARCHAR(4000) PATH '$.descr',
+        "$.DESCR" NVARCHAR(2147483647) FORMAT JSON PATH '$.descr',
+        ID INT PATH '$.ID',
+        "$.ID" NVARCHAR(2147483647) FORMAT JSON PATH '$.ID',
+        parent_ID INT PATH '$.parent_ID',
+        "$.PARENT_ID" NVARCHAR(2147483647) FORMAT JSON PATH '$.parent_ID',
+        "$.CHILDREN" NVARCHAR(2147483647) FORMAT JSON PATH '$.children'
+      )
+      ERROR ON ERROR
+    ) AS NEW
+  UNION ALL
+  SELECT
+    NEW.name,
+    NEW."$.NAME",
+    NEW.descr,
+    NEW."$.DESCR",
+    NEW.ID,
+    NEW."$.ID",
+    NEW.parent_ID,
+    NEW."$.PARENT_ID",
+    NEW."$.CHILDREN"
+  FROM 
+    JSON_TABLE(
+      :JSON,
+      '$.children[*].children[*]' COLUMNS(
+        name NVARCHAR(1020) PATH '$.name',
+        "$.NAME" NVARCHAR(2147483647) FORMAT JSON PATH '$.name',
+        descr NVARCHAR(4000) PATH '$.descr',
+        "$.DESCR" NVARCHAR(2147483647) FORMAT JSON PATH '$.descr',
+        ID INT PATH '$.ID',
+        "$.ID" NVARCHAR(2147483647) FORMAT JSON PATH '$.ID',
+        parent_ID INT PATH '$.parent_ID',
+        "$.PARENT_ID" NVARCHAR(2147483647) FORMAT JSON PATH '$.parent_ID',
+        "$.CHILDREN" NVARCHAR(2147483647) FORMAT JSON PATH '$.children'
+      )
+      ERROR ON ERROR
+    ) AS NEW;
+
+  -- DELETE all children of parents that are no longer in the dataset
```

Comment:
where are rows deleted which are not children of parents, but which are not part of the dataset?

E.g.

JSON input:

| ID  | parent_ID |
|-----|-----------|
| 1   | null      |
| 11  | 1         |
| 12  | 1         |
| 2   | null      |
| 21  | 2         |
| 22  | 2         |


Existing data:

| ID  | parent_ID | will be deleted? |
|-----|-----------|----------------|
| 13  | 1         |             yes |
| 3   | null      |             no |

Reply:
Note: They might be children of _other_ entities and would not have to be deleted.

Reply:
Thinking about it again, we can exclude non-children entities since compositions, by definition, are children and one needs to consider only children and children's children of the root instance.

We should also look at the case of a composition of one where the foreign key is part of the parent. Here, one cannot identify the previous child purely based on the payload and one needs to take into account the foreign keys of existing data.